crates/hir-ty/src/diagnostics/expr.rs

```diff
@@ -235,6 +243,36 @@ impl ExprValidator {
         }
         pattern
     }
+
+    fn validate_block(&mut self, db: &dyn HirDatabase, expr: &Expr) {
+        let Expr::Block { statements, .. } = expr else { return };
+        let body = db.body(self.owner);
+        let pattern_arena = Arena::new();
```

Comment:
Likewise, lets make this a field so we can re-use the arena from previous iterations (together with the one for general match checking)

Reply:
Is it beneficial to reuse an arena when there are no references to the old items? I thought it might be similar logic to reusing a Vec after `.clear()`ing it, but there is no equivalent method on arenas.

Reply:
Clearing isn't too beneficial for an arena, as that will just drop the allocated chunks which includes capacity, unlike a vec where capacity remains. The lifetime issues I can see, given how the arena is used. So without some lifetime transmuting, not much we can do about that.

Reply:
I think it drops all but the last chunk, which means there is still some good amortizing of capacity compared to making a new one. Also lifetimes should be cleaner with the new code, it might work now.

Reply:
Actually it would make sense to intern the lowered patterns if many parts of r-a start to use them

Reply:
I'm actually trying to remove the arena entirely, if this turns out ok perf-wise

Reply:
Yeah, don't worry about reusing the arena, it's going away soon